### ISSUE TITLE
Reference of Input to "on_change" event?

### ISSUE THREAD

**Issue by niveK77pur** [MunifTanjim/nui.nvim#189]
There exists #63 which allows to pass a reference of the Menu to its `on_change` callback. I was wondering if the same could be possible for the Input's `on_change` event?

Inspired by #63, I tried doing something similar on [this line](https://github.com/MunifTanjim/nui.nvim/blob/82851af021d651bb6bf6da0991c79f4c529ff37e/lua/nui/input/init.lua#L104). However that did not seem to have worked at all. 

```diff
diff --git a/lua/nui/input/init.lua b/lua/nui/input/init.lua
index 3a42010..c671146 100644
--- a/lua/nui/input/init.lua
+++ b/lua/nui/input/init.lua
@@ -101,7 +101,7 @@ function Input:init(popup_options, options)
     props.on_change = function()
       local value_with_prompt = vim.api.nvim_buf_get_lines(self.bufnr, 0, 1, false)[1]
       local value = string.sub(value_with_prompt, self._.prompt:length() + 1)
-      options.on_change(value)
+      options.on_change(value, self)
     end
   end
 end
```

*Actually, with my example below, it spits out a big error after hitting one character, which the error scared me. After the error, the input box seems to be dead, not responding to input neither being able to be closed anymore.*

```
Error executing lua callback: ...site/pack/packer/start/nui.nvim/lua/nui/popup/border.lua:503: E523: Not allowed here
stack traceback:
    [C]: in function 'nvim_buf_delete'
    ...site/pack/packer/start/nui.nvim/lua/nui/popup/border.lua:503: in function 'unmount'
    ...m/site/pack/packer/start/nui.nvim/lua/nui/popup/init.lua:271: in function 'unmount'
    /tmp/samba/nui-test.lua:52: in function 'on_change'
    ...m/site/pack/packer/start/nui.nvim/lua/nui/input/init.lua:104: in function <...m/site/pack/packer/start/nui.nvim/lua/nui/input/init.lua:1 01>
```

---

I want to create a mapping where I enter a vim register as a parameter via the nui input box. I know that a vim register will only be a single character long, so I was trying to use the `on_change` callback to close itself after I insert a character.

Here is something along the lines I was hoping to achieve. I created a wrapper function for myself called `nuiInput` which you can find further down.

```lua
-- simply echo contents of register
vim.keymap.set('n', 'Q', function()
    nuiInput{
        text = 'Register',
        on_change = function(value, input)
            if #value > 0 then
                input:unmount()
            end
        end,
        on_submit = function(reg)
            print(("Contents of %s: %s"):format(reg, vim.fn.getreg(reg)))
        end,
    }
end)
```

My `nuiInput` helper function.

```lua
function nuiInput(params) -- {{{
    params.position  = params.position  or "50%"
    params.width     = params.width     or 30
    params.text      = params.text      or 'Input Box'
    params.padding   = params.padding   or {1,1,1,1}
    params.prompt    = params.prompt    or nil
    params.default   = params.default   or nil
    params.on_close  = params.on_close  or nil
    params.on_submit = params.on_submit or nil
    params.on_change = params.on_change or nil

    local input = require('nui.input')({
        position = params.position,
        size = {
            width = params.width,
        },
        border = {
            style = ({ "double", "none", "rounded", "shadow", "single", "solid" })[3],
            text = {
                top = params.text,
                top_align = ({ "left", "right", "center" })[3],
            },
            padding = {
                top    = params.padding[1],
                left   = params.padding[2],
                bottom = params.padding[3],
                right  = params.padding[4],

            },
        },
        win_options = {
            winhighlight = "Normal:Normal,FloatBorder:SpecialChar",
        },
    }, {
        prompt        = params.prompt,
        default_value = params.default,
        on_close      = params.on_close,
        on_submit     = params.on_submit,
        on_change     = params.on_change,
    })

    -- mount/open the component
    input:mount()
end -- }}}
```

**Comment by MunifTanjim** [comment]
Can you try doing the unmount inside `vim.schedule`?

```lua
vim.schedule(function()
  input:unmount()
end)
```

**Comment by niveK77pur** [comment]
Thanks for the reply! Putting it inside `vim.schedule` does get rid of the error! Now the input window properly closes after hitting one character.

However, it seems that the `on_submit` callback is not being executed. Not sure now if that is a bug I introduced or if the `:unmount()` simply does not trigger it.

**Comment by MunifTanjim** [comment]
`on_submit` is trigged when user acutally submits. Unmounting is different from it. For your use-case, you can call your `on_submit` function from inside your `on_change` function. You'll need to extract the `on_submit` function and then use it.

```lua
local on_submit = function(...)
  -- do stuffs
end

local on_change = function(...)
  -- do stuffs
  -- on_submit(value)
  -- input:unmount()
end
```

**Comment by niveK77pur** [comment]
Doing it your way works! The print message immediately disappears though, but that's okay. I can get it back with `:mess` (also my actual use case doesn't involve printing, so no big deal).

I have actually tried something else though which seems to be and behave much cleaner. Basically, I use `nvim_feedkeys` to have the `on_change` callback hit a carriage return. This will then close the window and also trigger the `on_submit` callback!

I did it as follows, by modifying the `on_change` callback &mdash; instead of `:unmount()`, I made vim send a carriage return. This also won't require any changes to add a reference like Menu.
```lua
on_change = function(value)
    if #value > 0 then
        vim.api.nvim_feedkeys('\013', 'nt', false)
    end
end,
```

---

This basically solves my initial problem, I can't really think of any other situation where you might want to use a reference to the Input. I'll go ahead and close the issue! Thanks a bunch for the inputs!
